src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala

```diff
@@ -178,16 +180,60 @@ object ZipAndJarSourcePathFactory extends ZipAndJarFileLookupFactory {
 
 final class FileBasedCache[T] {
   import java.nio.file.Path
-  private case class Stamp(lastModified: FileTime, fileKey: Object)
+  private case class Stamp(size: Long, signatures: ArrayBuffer[Long], fileKey: Object)
   private val cache = collection.mutable.Map.empty[Seq[Path], (Seq[Stamp], T)]
 
+  /**
+    * Read the CRCs of every zip entry if all of them succeed to be read.
+    *
+    * This method uses `java.util.jar` instead of `java.util.zip` utilities
+    * because the method `getCrc` in [[java.util.zip.ZipEntry]] returns 0 or
+    * -1 values.
+    *
+    * @param jarFile A jar file (which follows the zip specification format).
+    * @return A list of checksums for every zip entry.
+    */
+  def readJarCrcs(jarFile: Path): Option[ArrayBuffer[Long]] = {
+    try {
+      val jf = new java.util.jar.JarFile(jarFile.toFile)
+      try {
+        val entries = jf.entries()
+        val buffer = new ArrayBuffer[Long](32)
+        while (entries.hasMoreElements) buffer.+=(entries.nextElement().getCrc)
```

Comment:
this signature will not detect renames, as the resource name is not considered

This may be safe for class files (as their content will change if the name changes) and for java source files, but not safe for scala source files which change directory for instance. The package is still the same, the content is the same, and the CRC is the same, but the mapping has changed, and could provide surprising results IMO

I think it would be better to include the entry name hash as well

Reply:
> but not safe for scala source files which change directory for instance. 
  
Do you have an example of a change in a scala source file that is relevant for compilation and *doesn't cause* a change in the bytecode? Because it seems like a stretch.

Reply:
As I said earlier this is not a problem for class files, therefore I am not considering bytecode. 

This code path is also used to sourcepath as I understand it. If I move a scala file from one directory to another (or rename a directory) without changing the content the files content (and therefore the CRC) will not change, but the entry name will. 
Scala doesn't mandate that the package matches the path, so this is legal

Lookups are however path based so the results are not the same

Reply:
> This code path is also used to sourcepath as I understand it.
  
Is it? My understanding is that it's only used for the classpath.

Reply:
My mistake - its not for sources
```
object ZipAndJarSourcePathFactory extends ZipAndJarFileLookupFactory {
```
there is a parallel factory for sources

**The problem still exists though**. A macro or plugin can do what it wants, so it could read a resource from its classpath (which could have changed name), or an entry from the jar manifest (which I presume is covered, if it is not erased CRC) 

This is not an issue for code that doesn't actually run (i.e. the compile classpath) though. I don't think that we gain much from this distinction though, unless there are many build systems that package source or other artifacts in their code jars.